src/modules/wolfssl/nm_wolfssl_cli.c

```diff
@@ -0,0 +1,727 @@
+    wolfSSL_CTX_use_certificate_buffer(ctx->ctx, certificate, certificateSize, WOLFSSL_FILETYPE_PEM) != WOLFSSL_SUCCESS) {
+        NABTO_LOG_ERROR(LOG, "wolfSSL_CTX_use_certificate_buffer");
+        return NABTO_EC_UNKNOWN;
+    }
+
+    return NABTO_EC_OK;
+}
+
+np_error_code nm_wolfssl_cli_set_root_certs(struct np_dtls_cli_context* ctx, const char* rootCerts)
+{
+    if (wolfSSL_CTX_load_verify_buffer(ctx->ctx, (const unsigned char*)rootCerts, strlen(rootCerts), WOLFSSL_FILETYPE_PEM) != WOLFSSL_SUCCESS)
+    {
+        NABTO_LOG_ERROR(LOG, "cannot load ca certificate");
+        return NABTO_EC_FAILED;
+    }
+    return NABTO_EC_OK;
+}
+
+np_error_code nm_wolfssl_cli_disable_certificate_validation(struct np_dtls_cli_context* ctx)
+{
+    wolfSSL_set_verify(ctx->ssl, (SSL_VERIFY_NONE), NULL);
+    return NABTO_EC_OK;
+}
+
+/**
+ * get peers fingerprint for given DTLS client context
+ */
+np_error_code get_fingerprint(struct np_dtls_cli_context* ctx, uint8_t* fp)
+{
+    // Get client fingerprint
+    WOLFSSL_X509 *crt = wolfSSL_get_peer_certificate(ctx->ssl);
+    if (!crt) {
+        return NABTO_EC_UNKNOWN;
+    }
+    return nm_wolfssl_util_fp_from_crt(crt, fp);
+}
+
+np_error_code set_handshake_timeout(struct np_dtls_cli_context* ctx, uint32_t minTimeout, uint32_t maxTimeout)
+{
+    int min = minTimeout / 1000;
+    int max = maxTimeout / 1000;
+    if (wolfSSL_dtls_set_timeout_init(ctx->ssl, min) !=
+        WOLFSSL_SUCCESS) {
+        NABTO_LOG_ERROR(LOG, "Cannot set min timeout for DTLS client connection");
+        return NABTO_EC_FAILED;
+    };
+    if (wolfSSL_dtls_set_timeout_max(ctx->ssl, max) != WOLFSSL_SUCCESS) {
+        NABTO_LOG_ERROR(LOG, "Cannot set max timeout for DTLS client connection");
+        return NABTO_EC_FAILED;
+    }
+    return NABTO_EC_OK;
+}
+
+/*
+ * asyncroniously start a dtls connection
+ */
+np_error_code nm_dtls_connect(struct np_dtls_cli_context* ctx)
+{
+    ctx->state = CONNECTING;
+
+    nm_dtls_event_do_one(ctx);
+    return NABTO_EC_OK;
+}
+
+/*
+ * Handle events for the connection phase
+ */
+void nm_dtls_event_do_one(void* data)
+{
+    struct np_dtls_cli_context* ctx = data;
+    int ret;
+    if(ctx->state == CONNECTING) {
+        ret = wolfSSL_connect(ctx->ssl);
+        if (ret != WOLFSSL_SUCCESS) {
+            int err = wolfSSL_get_error(ctx->ssl, ret);
+            if (err == WOLFSSL_ERROR_WANT_READ) {
+                NABTO_LOG_TRACE(LOG, "Want Read");
+                set_timeout(ctx);
+            } else if (err == WOLFSSL_ERROR_WANT_WRITE){
+                NABTO_LOG_TRACE(LOG, "Want Write");
+                // Wait for IO to happen
+            } else {
+                enum np_dtls_cli_event event = NP_DTLS_CLI_EVENT_CLOSED;
+                char buf[80];
+                wolfSSL_ERR_error_string(err, buf);
+                if (err > MIN_CODE_E) { // All wolfCrypt errors range ]0...MIN_CODE_E[
```

Comment:
Make a comment about that we are maybe matching non certificate verifification errors.